File: ai_engine/ai_character.py
#!/usr/bin/env python3
"""
AI Character Module - Defines AI personalities and learning capabilities
Path: ai_engine/ai_character.py
"""

import json
import pickle
from pathlib import Path
from typing import Dict, List, Optional
from dataclasses import dataclass, field


@dataclass
class PersonalityTraits:
    """AI personality characteristics"""
    aggression: float = 0.5      # 0.0 - 1.0
    defensiveness: float = 0.5
    risk_taking: float = 0.5
    patience: float = 0.5
    tactical: float = 0.5
    positional: float = 0.5
    
    def to_dict(self) -> Dict:
        return {
            'aggression': self.aggression,
            'defensiveness': self.defensiveness,
            'risk_taking': self.risk_taking,
            'patience': self.patience,
            'tactical': self.tactical,
            'positional': self.positional
        }


@dataclass
class PerformanceMetrics:
    """Track AI performance statistics"""
    games_played: int = 0
    games_won: int = 0
    games_lost: int = 0
    games_drawn: int = 0
    brilliant_moves: int = 0
    good_moves: int = 0
    mistakes: int = 0
    blunders: int = 0
    average_accuracy: float = 0.0
    
    @property
    def win_rate(self) -> float:
        if self.games_played == 0:
            return 0.0
        return (self.games_won / self.games_played) * 100


class AICharacter:
    """Represents an AI character with learning capabilities"""
    
    def __init__(self, name: str, elo_rating: int = 1000):
        self.name = name
        self.elo_rating = elo_rating
        self.personality = PersonalityTraits()
        self.metrics = PerformanceMetrics()
        
        # Learning data
        self.opening_preferences: Dict[str, float] = {}
        self.move_patterns: Dict[str, int] = {}
        self.favorite_openings: List[str] = []
        
    def update_elo(self, opponent_elo: int, result: str):
        """Update ELO rating based on game result"""
        K = 32  # K-factor
        
        expected = 1.0 / (1.0 + 10 ** ((opponent_elo - self.elo_rating) / 400.0))
        
        actual = {'win': 1.0, 'draw': 0.5, 'loss': 0.0}.get(result.lower(), 0.5)
        
        delta = K * (actual - expected)
        self.elo_rating += int(delta)
        
        # Update metrics
        self.metrics.games_played += 1
        if result.lower() == 'win':
            self.metrics.games_won += 1
        elif result.lower() == 'loss':
            self.metrics.games_lost += 1
        else:
            self.metrics.games_drawn += 1
    
    def learn_from_move(self, move_uci: str, evaluation: float, quality: str):
        """Learn from a move based on its quality"""
        self.move_patterns[move_uci] = self.move_patterns.get(move_uci, 0) + 1
        
        # Update metrics
        if quality == 'brilliant':
            self.metrics.brilliant_moves += 1
            self.personality.tactical = min(1.0, self.personality.tactical + 0.01)
        elif quality == 'good':
            self.metrics.good_moves += 1
        elif quality == 'mistake':
            self.metrics.mistakes += 1
            self.personality.patience = min(1.0, self.personality.patience + 0.01)
        elif quality == 'blunder':
            self.metrics.blunders += 1
            self.personality.risk_taking = max(0.0, self.personality.risk_taking - 0.02)
    
    def learn_opening(self, opening_name: str, success: bool):
        """Learn from opening performance"""
        current = self.opening_preferences.get(opening_name, 0.5)
        
        if success:
            self.opening_preferences[opening_name] = min(1.0, current + 0.05)
        else:
            self.opening_preferences[opening_name] = max(0.0, current - 0.03)
        
        self._update_favorite_openings()
    
    def _update_favorite_openings(self):
        """Update list of favorite openings"""
        sorted_openings = sorted(
            self.opening_preferences.items(),
            key=lambda x: x[1],
            reverse=True
        )
        self.favorite_openings = [name for name, _ in sorted_openings[:5]]
    
    def get_playing_style(self) -> str:
        """Get description of playing style"""
        traits = self.personality
        
        if traits.aggression > 0.7:
            return "Aggressive"
        elif traits.defensiveness > 0.7:
            return "Defensive"
        elif traits.tactical > 0.7:
            return "Tactical"
        elif traits.positional > 0.7:
            return "Positional"
        elif traits.patience > 0.7:
            return "Patient"
        else:
            return "Balanced"
    
    def get_strength_level(self) -> str:
        """Get strength level based on ELO"""
        if self.elo_rating >= 2400:
            return "Grandmaster"
        elif self.elo_rating >= 2200:
            return "International Master"
        elif self.elo_rating >= 2000:
            return "Expert"
        elif self.elo_rating >= 1800:
            return "Advanced"
        elif self.elo_rating >= 1600:
            return "Intermediate"
        elif self.elo_rating >= 1400:
            return "Developing"
        elif self.elo_rating >= 1200:
            return "Beginner"
        else:
            return "Novice"
    
    def save(self, directory: Path):
        """Save character to file"""
        directory.mkdir(parents=True, exist_ok=True)
        filepath = directory / f"{self.name}.pkl"
        
        with open(filepath, 'wb') as f:
            pickle.dump(self, f)
    
    @classmethod
    def load(cls, filepath: Path) -> 'AICharacter':
        """Load character from file"""
        with open(filepath, 'rb') as f:
            return pickle.load(f)
    
    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization"""
        return {
            'name': self.name,
            'elo_rating': self.elo_rating,
            'personality': self.personality.to_dict(),
            'metrics': {
                'games_played': self.metrics.games_played,
                'games_won': self.metrics.games_won,
                'games_lost': self.metrics.games_lost,
                'games_drawn': self.metrics.games_drawn,
                'win_rate': self.metrics.win_rate,
                'brilliant_moves': self.metrics.brilliant_moves,
                'mistakes': self.metrics.mistakes,
                'blunders': self.metrics.blunders
            },
            'playing_style': self.get_playing_style(),
            'strength_level': self.get_strength_level(),
            'favorite_openings': self.favorite_openings
        }
    
    def __str__(self) -> str:
        return (f"{self.name} (ELO: {self.elo_rating}, "
                f"Style: {self.get_playing_style()}, "
                f"Level: {self.get_strength_level()})")

File: src/main/java/com/erland/chess/ai/AICharacter.java
package com.erland.chess.ai;

import java.io.Serializable;
import java.util.*;

/**
 * Representasi karakter AI yang dapat belajar
 */
public class AICharacter implements Serializable {
    private static final long serialVersionUID = 1L;
    
    private String name;
    private int eloRating;
    private int gamesPlayed;
    private int gamesWon;
    private int gamesLost;
    private int gamesDraw;
    
    // Personality traits (0.0 - 1.0)
    private double aggression;      // Seberapa agresif (suka menyerang)
    private double defensiveness;   // Seberapa defensif
    private double riskTaking;      // Seberapa berani ambil resiko
    private double patience;        // Seberapa sabar (endgame skill)
    private double tactical;        // Kemampuan taktik
    private double positional;      // Pemahaman posisi
    
    // Learning data
    private Map<String, Double> openingPreferences = new HashMap<>();
    private Map<String, Integer> movePatterns = new HashMap<>();
    private List<String> favoriteOpenings = new ArrayList<>();
    
    // Performance metrics
    private double averageAccuracy = 0.0;
    private int brilliantMoves = 0;
    private int blunders = 0;
    
    public AICharacter(String name) {
        this(name, 1000); // Default ELO
    }
    
    public AICharacter(String name, int initialElo) {
        this.name = name;
        this.eloRating = initialElo;
        this.gamesPlayed = 0;
        this.gamesWon = 0;
        this.gamesLost = 0;
        this.gamesDraw = 0;
        
        // Random personality
        Random rand = new Random();
        this.aggression = 0.3 + rand.nextDouble() * 0.4;
        this.defensiveness = 0.3 + rand.nextDouble() * 0.4;
        this.riskTaking = 0.2 + rand.nextDouble() * 0.6;
        this.patience = 0.3 + rand.nextDouble() * 0.4;
        this.tactical = 0.3 + rand.nextDouble() * 0.4;
        this.positional = 0.3 + rand.nextDouble() * 0.4;
    }
    
    /**
     * Update ELO rating based on game result
     */
    public void updateElo(int opponentElo, GameResult result) {
        int K = 32; // K-factor for ELO calculation
        
        double expectedScore = 1.0 / (1.0 + Math.pow(10, (opponentElo - eloRating) / 400.0));
        
        double actualScore = 0.0;
        switch (result) {
            case WIN:
                actualScore = 1.0;
                gamesWon++;
                break;
            case DRAW:
                actualScore = 0.5;
                gamesDraw++;
                break;
            case LOSS:
                actualScore = 0.0;
                gamesLost++;
                break;
        }
        
        gamesPlayed++;
        int eloDelta = (int) (K * (actualScore - expectedScore));
        eloRating += eloDelta;
        
        System.out.println(name + " ELO: " + eloRating + " (" + (eloDelta >= 0 ? "+" : "") + eloDelta + ")");
    }
    
    /**
     * Learn from a move
     */
    public void learnMove(String moveUCI, double evaluation, String quality) {
        // Update move patterns
        movePatterns.put(moveUCI, movePatterns.getOrDefault(moveUCI, 0) + 1);
        
        // Update personality based on move quality
        if ("brilliant".equals(quality)) {
            brilliantMoves++;
            // Reinforce the traits that led to this move
            if (evaluation > 0) {
                aggression = Math.min(1.0, aggression + 0.01);
                tactical = Math.min(1.0, tactical + 0.01);
            }
        } else if ("blunder".equals(quality)) {
            blunders++;
            // Adjust traits to avoid similar mistakes
            riskTaking = Math.max(0.0, riskTaking - 0.01);
            patience = Math.min(1.0, patience + 0.01);
        }
    }
    
    /**
     * Learn from opening
     */
    public void learnOpening(String opening, boolean successful) {
        double current = openingPreferences.getOrDefault(opening, 0.5);
        if (successful) {
            openingPreferences.put(opening, Math.min(1.0, current + 0.05));
        } else {
            openingPreferences.put(opening, Math.max(0.0, current - 0.03));
        }
        
        // Update favorite openings
        updateFavoriteOpenings();
    }
    
    private void updateFavoriteOpenings() {
        favoriteOpenings.clear();
        openingPreferences.entrySet().stream()
            .sorted(Map.Entry.<String, Double>comparingByValue().reversed())
            .limit(5)
            .forEach(e -> favoriteOpenings.add(e.getKey()));
    }
    
    /**
     * Get playing style description
     */
    public String getPlayingStyle() {
        if (aggression > 0.6) return "Aggressive";
        if (defensiveness > 0.6) return "Defensive";
        if (tactical > 0.6) return "Tactical";
        if (positional > 0.6) return "Positional";
        if (patience > 0.6) return "Patient";
        return "Balanced";
    }
    
    /**
     * Get strength level based on ELO
     */
    public String getStrengthLevel() {
        if (eloRating >= 2400) return "Grandmaster";
        if (eloRating >= 2200) return "International Master";
        if (eloRating >= 2000) return "Expert";
        if (eloRating >= 1800) return "Advanced";
        if (eloRating >= 1600) return "Intermediate";
        if (eloRating >= 1400) return "Developing";
        if (eloRating >= 1200) return "Beginner";
        return "Novice";
    }
    
    /**
     * Get win rate
     */
    public double getWinRate() {
        if (gamesPlayed == 0) return 0.0;
        return (double) gamesWon / gamesPlayed * 100;
    }
    
    // Getters and Setters
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public int getEloRating() { return eloRating; }
    public void setEloRating(int eloRating) { this.eloRating = eloRating; }
    
    public int getGamesPlayed() { return gamesPlayed; }
    public int getGamesWon() { return gamesWon; }
    public int getGamesLost() { return gamesLost; }
    public int getGamesDraw() { return gamesDraw; }
    
    public double getAggression() { return aggression; }
    public double getDefensiveness() { return defensiveness; }
    public double getRiskTaking() { return riskTaking; }
    public double getPatience() { return patience; }
    public double getTactical() { return tactical; }
    public double getPositional() { return positional; }
    
    public double getAverageAccuracy() { return averageAccuracy; }
    public void setAverageAccuracy(double accuracy) { this.averageAccuracy = accuracy; }
    
    public int getBrilliantMoves() { return brilliantMoves; }
    public int getBlunders() { return blunders; }
    
    public List<String> getFavoriteOpenings() { return new ArrayList<>(favoriteOpenings); }
    
    public enum GameResult {
        WIN, DRAW, LOSS
    }
    
    @Override
    public String toString() {
        return String.format("%s (ELO: %d, Style: %s, Level: %s)", 
            name, eloRating, getPlayingStyle(), getStrengthLevel());
    }
}